core/kotlinx-coroutines-core/src/Executors.kt

```diff
@@ -10,43 +10,37 @@ import java.util.concurrent.*
 import kotlin.coroutines.experimental.*
 
 /**
- * [CoroutineDispatcher] that implements [Closeable]
+ * [CoroutineDispatcher] that has underlying [Executor] for dispatching tasks.
+ * Instances of [ExecutorCoroutineDispatcher] should be closed by the owner of the dispatcher.
+ *
+ * This class is generally used as a bridge between coroutine-based API and
+ * asynchronous API which requires instance of the [Executor]
  */
-abstract class CloseableCoroutineDispatcher: CoroutineDispatcher(), Closeable
+public abstract class ExecutorCoroutineDispatcher: CoroutineDispatcher(), Closeable {
 
-/**
- * Converts an instance of [ExecutorService] to an implementation of [CloseableCoroutineDispatcher].
- */
-public fun ExecutorService.asCoroutineDispatcher(): CloseableCoroutineDispatcher =
-    // we know that an implementation of Executor.asCoroutineDispatcher actually returns a closeable one
-    (this as Executor).asCoroutineDispatcher() as CloseableCoroutineDispatcher
+    /**
+     * Underlying executor of current [CoroutineDispatcher].
+     */
+    abstract val executor: Executor
+}
 
 /**
- * Converts an instance of [Executor] to an implementation of [CoroutineDispatcher].
- * @suppress **Deprecated**: Renamed to [asCoroutineDispatcher].
+ * Converts an instance of [ExecutorService] to an implementation of [ExecutorCoroutineDispatcher].
  */
-@Deprecated("Renamed to `asCoroutineDispatcher`",
-    replaceWith = ReplaceWith("asCoroutineDispatcher()"))
-public fun Executor.toCoroutineDispatcher(): CoroutineDispatcher =
-    ExecutorCoroutineDispatcher(this)
+public fun ExecutorService.asCoroutineDispatcher(): ExecutorCoroutineDispatcher =
+    // we know that an implementation of Executor.asCoroutineDispatcher actually returns a closeable one
+    (this as Executor).asCoroutineDispatcher() as ExecutorCoroutineDispatcher
 
 /**
  * Converts an instance of [Executor] to an implementation of [CoroutineDispatcher].
  */
 public fun Executor.asCoroutineDispatcher(): CoroutineDispatcher =
-    ExecutorCoroutineDispatcher(this)
+    ExecutorCoroutineDispatcherImpl(this)
 
-private class ExecutorCoroutineDispatcher(override val executor: Executor) : ExecutorCoroutineDispatcherBase()
+private class ExecutorCoroutineDispatcherImpl(override val executor: Executor) : ExecutorCoroutineDispatcherBase()
+
+abstract class ExecutorCoroutineDispatcherBase : ExecutorCoroutineDispatcher(), Delay {
```

Comment:
This abstract class shall is public will be visible in generated documentation. It shall be either fully documented with KDoc or suppressed from documentation (as it was before).